crates/typst-kit/src/package.rs

```diff
@@ -14,6 +15,111 @@ use typst::syntax::package::{
+                .open(&lock_file_path)
+                .map_err(other_err("failed to open lock file"))?,
+            path: lock_file_path,
+            tmp_path: format!("{parent_dir_path}/{lock_file_name}"),
+        })
+    }
+
+    /// Try acquiring exclusive lock.
+    fn try_lock(&self) -> io::Result<()> {
+        self.file.try_lock_exclusive()
+    }
+
+    /// Acquire exclusive lock.
+    fn lock(&self) -> Result<(), PackageError> {
+        self.file
+            .lock_exclusive()
+            .map_err(other_err("failed to aquire lock file"))
+    }
+
+    /// Try acquiring the exclusive lock, if failed print an error an wait
+    /// for the lock with blocking.
+    ///
+    /// Returns `Ok(true)` if acquired first try, otherwise `Ok(false)`.
+    fn try_and_lock(&self) -> Result<bool, PackageError> {
+        match self.try_lock() {
+            Ok(_) => {
+                eprintln!("[tmp] acquired the lock file {:?}", self.tmp_path);
+                Ok(true)
+            }
+            Err(ref err) if err.kind() == io::ErrorKind::WouldBlock => {
+                eprintln!("[tmp] couldn't aquire the lock file {:?}", self.tmp_path);
+                eprintln!(
+                    "Waiting for another instance to finish installing the package..."
+                );
+                self.lock()?;
+
+                // // If other instance successfully installed a package after
+                // // waiting for the lock, then there is nothing left to do.
+                // if dir.exists() {
+                //     eprintln!("[tmp] dropped the lock file {:?}", self.tmp_path);
+                //     return Ok(dir);
+                // }
+
+                // eprintln!(
+                //     "Another instance failed to install the package, trying it again."
+                // );
+                Ok(false)
+                // Additional cool/user-friendly logs:
+                // let _write_lock = lock_file.write().unwrap();
+                // println!("checking if package already exists");
+                // if dir.exists() {
+                //     println!("it does! no need to download and unpack");
+                //     return Ok(dir);
+                // } else {
+                //     println!("waited to aquire the file lock and the other instance wasn't able to download it :/");
+                //     todo!()
+                // }
+            }
+            Err(err) => Err(other_err("failed to aquire lock file")(err)),
+        }
+    }
+
+    /// Remove (delete) lock file (if wasn't already by other process).
+    ///
+    /// Will be automatically called when struct is being dropped ([Drop]).
+    fn remove(&self) -> Result<(), PackageError> {
+        if self.path.exists() {
+            fs::remove_file(&self.path)
+                .map_err(other_err("failed to remove lock file"))?;
+        }
+        eprintln!("[tmp] dropped the lock file {:?}", self.tmp_path);
+        Ok(())
+    }
+}
```

Comment:
Some of these should be public.

Reply:
We should be careful to not over-engineer this - we just need to expose an API to lock what's relevant to us (that is, the package cache), and abstract away the implementation details.

I feel like Gleam's API for this (https://github.com/gleam-lang/gleam/blob/f9c2d98970241e51fba58cf998e10245f8a2fe05/compiler-cli/src/build_lock.rs) is a great reference for how to proceed in this aspect as well.

Reply:
Even though 1-method usage sounds very convenient, it doesn't feel idiomatic.

Reply:
I've created `create_and_lock()` method. Additionally, I've kept the `true`/`false` value when checking if the lock was acquired right away. So, if you need to do some additional checks (yes) or messages (yes?), then you can call it. That's it. After going out of scope, the lock file will be removed.

Reply:
This can potentially be changed to `LockFile::create_and_lock(..., |1st_try| { if 1st_try { ... }; ...; })`, which will give a clear scope, but maybe potential borrowing errors for beginners? This will also remove the need for `Drop` implementation.